File: DepartamentoPessoal.Server/Services/UsuarioService.cs
using CommonTemp.DTOs.Inputs.Usuario;
using CommonTemp.DTOs.Outputs.Usuario;
using CommonTemp.Models;
using DepartamentoPessoal.Server.Data;
using DepartamentoPessoal.Server.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace DepartamentoPessoal.Server.Services;

public class UsuarioService(AppDbContext db) : IUsuarioService
{
    public async Task<IEnumerable<UsuarioOutput>> GetAllAsync()
    {
        return await db.Usuarios
            .AsNoTracking()
            .OrderBy(u => u.Nome)
            .Select(u => MapToOutput(u))
            .ToListAsync();
    }

    public async Task<UsuarioOutput?> GetByIdAsync(int id)
    {
        var u = await db.Usuarios.AsNoTracking().FirstOrDefaultAsync(x => x.Id == id);
        return u is null ? null : MapToOutput(u);
    }

    public async Task<UsuarioOutput> CreateAsync(UsuarioInput input)
    {
        var usuario = new Usuario
        {
            Nome = input.Nome,
            Email = input.Email.ToLower().Trim(),
            SenhaHash = BCrypt.Net.BCrypt.HashPassword(input.Senha),
            Perfil = input.Perfil,
            Ativo = true
        };

        db.Usuarios.Add(usuario);
        await db.SaveChangesAsync();
        return MapToOutput(usuario);
    }

    public async Task<UsuarioOutput?> UpdateAsync(int id, UsuarioInput input)
    {
        var usuario = await db.Usuarios.FindAsync(id);
        if (usuario is null) return null;

        usuario.Nome = input.Nome;
        usuario.Email = input.Email.ToLower().Trim();
        usuario.Perfil = input.Perfil;

        if (!string.IsNullOrWhiteSpace(input.Senha))
            usuario.SenhaHash = BCrypt.Net.BCrypt.HashPassword(input.Senha);

        await db.SaveChangesAsync();
        return MapToOutput(usuario);
    }

    public async Task<bool> AlterarSenhaAsync(int id, AlterarSenhaInput input)
    {
        var usuario = await db.Usuarios.FindAsync(id);
        if (usuario is null) return false;
        if (!BCrypt.Net.BCrypt.Verify(input.SenhaAtual, usuario.SenhaHash)) return false;

        usuario.SenhaHash = BCrypt.Net.BCrypt.HashPassword(input.NovaSenha);
        await db.SaveChangesAsync();
        return true;
    }

    public async Task<bool> ToggleAtivoAsync(int id)
    {
        var usuario = await db.Usuarios.FindAsync(id);
        if (usuario is null) return false;
        usuario.Ativo = !usuario.Ativo;
        await db.SaveChangesAsync();
        return true;
    }

    public async Task<bool> DeleteAsync(int id)
    {
        var usuario = await db.Usuarios.FindAsync(id);
        if (usuario is null) return false;
        db.Usuarios.Remove(usuario);
        await db.SaveChangesAsync();
        return true;
    }

    private static UsuarioOutput MapToOutput(Usuario u) => new()
    {
        Id = u.Id,
        Nome = u.Nome,
        Email = u.Email,
        Perfil = u.Perfil,
        PerfilDescricao = u.Perfil.ToString(),
        Ativo = u.Ativo,
        CriadoEm = u.CriadoEm,
        UltimoAcessoEm = u.UltimoAcessoEm
    };
}

File: departamentopessoal.client/src/services/usuarioService.ts
import api from './api'

export interface UsuarioOutput {
  id: number
  nome: string
  email: string
  perfil: string
  perfilDescricao: string
  ativo: boolean
  colaboradorId?: number
  colaboradorNome?: string
  criadoEm: string
  ultimoAcessoEm: string | null
}

export interface UsuarioInput {
  nome: string
  email: string
  senha: string
  perfil: string
  colaboradorId?: number
}

export interface AlterarSenhaInput {
  senhaAtual: string
  novaSenha: string
}

export const usuarioService = {
  getAll: () => api.get<UsuarioOutput[]>('/usuarios').then(r => r.data),
  getById: (id: number) => api.get<UsuarioOutput>(`/usuarios/${id}`).then(r => r.data),
  create: (data: UsuarioInput) => api.post<UsuarioOutput>('/usuarios', data).then(r => r.data),
  update: (id: number, data: UsuarioInput) => api.put<UsuarioOutput>(`/usuarios/${id}`, data).then(r => r.data),
  toggleAtivo: (id: number) => api.patch<UsuarioOutput>(`/usuarios/${id}/toggle-ativo`).then(r => r.data),
  alterarSenha: (id: number, data: AlterarSenhaInput) => api.patch(`/usuarios/${id}/senha`, data),
  delete: (id: number) => api.delete(`/usuarios/${id}`),
}
